feat: unstoppable domains upgrade + reverse resolution

amritkumarj opened the pull request:
### Description
Upgraded unstoppable domain package and added reverse resolution 

--- 

[umbra integration .webm](https://user-images.githubusercontent.com/72156537/184285064-7d1b3f4e-2c9f-4c50-af9d-1fcd55b42e4b.webm)
 

Reverse Resolution : 
![reverse resolution 1](https://user-images.githubusercontent.com/72156537/184285318-02a6ac9b-cf82-48d3-b15b-e1565526e5f6.jpg)
![reverse resolution 2](https://user-images.githubusercontent.com/72156537/184285321-ace3acf5-e7b6-45c2-a5ec-d751cc760d0e.jpg)

--- 

You can also use our Unstoppable Domains to check lookup Functionality : blockdudes.blockchain resolved to 0x89f9798a8d8bac15a47031779cc6aa08d8059c52

apbendi reviewed (commented):
Hey @amritkumarj, overall things seem to be working, but here are a couple things I noticed on my QA pass:

* Running `yarn` locally results in a diff to the `yarn.lock` file. We should check in the latest lockfile
* There's a regression on the Receive screen table. Clicking on an address copies the address. If that address is a domain name, it should copy the *address*, not the domain name shown. You can test this out in current production.
* If a user tries to withdraw to their own address, they should see a warning that says "It resolves to the same address as the connected wallet". Currently that does not show up for CNS names (see screenshots).

<img width="583" alt="Screen Shot 2022-09-08 at 10 42 46 PM" src="https://user-images.githubusercontent.com/1479802/189262254-a5511bdf-69a1-48ab-a3b7-51489fae83f1.png">

<img width="590" alt="Screen Shot 2022-09-08 at 10 40 57 PM" src="https://user-images.githubusercontent.com/1479802/189262260-3e27083d-2dbe-4446-aa79-402f830127dc.png">

Also, do you mind rebasing this to the latest `master`? Thanks for your patience in our review.

amritkumarj commented:
@apbendi Thanks for the review

1. Fixed
2. The issue is same for the ENS address, but I added this functionality so it is also solved
3. The issue was the casing, so I lowercased both address at the time of checking

apbendi commented:
thanks @amritkumarj

> The issue is same for the ENS address, but I added this functionality so it is also solved

Right it broken on ENS for this PR but wasn't in prod! Seems to be working for both case now 👍 

> The issue was the casing, so I lowercased both address at the time of checking, is that ok?

You mean you downcase both addresses to do the compare, right? If so, sounds perfectly reasonable.

Overall PR is looking good and seems to work well! Nice job. One annoying request: instead of the merge commit of master into your branch, can you remove the merge commit and rebase the branch instead? We try to keep merges out of history.

@mds1 could you give the code another quick look over as well? From my perspective looks good, but it's touching a lot of sections you know best. Thanks!

mds1 commented on frontend/src/utils/address.ts:
```diff
@@ -43,20 +44,9 @@ export const lookupEnsName = async (address: string, provider: Provider) => {
-    const names = json.data.domains;
-    return names.length > 0 ? names[0].name : null;
+    const resolution = new Resolution();
+    const domain = await resolution.reverse(address);
```
Can you explain how forward and resolution should work across chains? For example, here's what we do for ENS:
- If using mainnet, forward and reverse resolution is done using mainnet contracts. Reverse resolution (from address to name) is validated by ensuring the forward lookup for the name matches.
- If rinkeby, same as mainnet, but using rinkeby contracts.
- For all other networks, use the mainnet resolution flow. If withdrawing to an ENS name, and it resolves to a contract, warn that funds may not be accessible on the current network

amritkumarj replied:
Yeah, the implementation is same as ENS

mds1 commented on umbra-js/src/utils/utils.ts:
```diff
@@ -316,12 +306,11 @@ async function resolveEns(name: string, provider: EthersProvider) {
   try {
-    const resolution = getResolutionInstance(provider);
+    const resolution = getResolutionInstance();
     const address = await resolution.addr(name, 'ETH');
```
Does this return a checksum address? We try to use with checksummed addresses everywhere

amritkumarj replied:
I don't think so, I am now converting this address into checksum address using getAddress function of ethersjs

![image](https://user-images.githubusercontent.com/72156537/191677872-d4a8ff92-ce59-4c1c-89b5-eb0cf402fc12.png)

mds1 replied:
Nice that looks perfect! (Though I don't think you pushed that change)

amritkumarj commented on umbra-js/src/utils/utils.ts:
```diff
@@ -276,19 +276,9 @@ export function isDomain(name: string) {
-      },
-    },
-  });
+function getResolutionInstance() {
```
In the new version of Unstoppable Domains we still use RPC calls which is inbuilt in the library, as you can check here https://github.com/unstoppabledomains/resolution/blob/master/src/UnsInternal.ts

![image](https://user-images.githubusercontent.com/72156537/191676580-61034cd5-9e67-4152-8957-98f220e010a6.png)

mds1 replied:
Ok great, we can start with this and if requests get throttled or fail due to rate limits from shared keys, it seems we can pass in our own provider to the constructor using `source.provider`?

amritkumarj commented on frontend/src/utils/address.ts:
```diff
@@ -43,20 +44,9 @@ export const lookupEnsName = async (address: string, provider: Provider) => {
-    const json = (await res.json()) as CnsQueryResponse;
-    const names = json.data.domains;
-    return names.length > 0 ? names[0].name : null;
+    const resolution = new Resolution();
```
Same uns library have inbuilt RPC providers

amritkumarj commented on yarn.lock:
```diff
@@ -5030,18 +5030,16 @@
-  version "3.0.0"
-  resolved "https://registry.yarnpkg.com/@unstoppabledomains/resolution/-/resolution-3.0.0.tgz#148080757f3ab2e94993228a5934ea60f1e0807b"
-  integrity sha512-iAvVkDWsSdi9WOiCp8t6GwWQZYIQjca+JAzfnNB7LvRDim3jrgcLlbeSYu+Mwzn2JKUGf/Ygw7YtnmjihnT5fw==
+"@unstoppabledomains/resolution@8.3.1":
```
So, here the tests are failing because the stealth keys are added in the StealthKeyRegistry of Umbra not Unstoppable Domains. 
Should I remove these tests or change the test to use StealthKeyRegistry of Umbra or any other solution you have?

mds1 replied:
Ah, sorry, yes storing the keys directly on the name is now deprecated. We used to do that for both ENS and CNS, but for various reasons we migrated everything to the `StealthKeyRegistry`. So now we take an ENS/CNS name, resolve it to an address, and look up that address in the registry contract.

For the tests, we hardcode a name (previously it was `udtestdev-msolomon.crypto`, but we may need a new one now?) which has its pubkeys set on Rinkeby, and run the test against rinkeby (it's read-only, no txs are sent). We also are not forced to use rinkeby, since it's being deprecated anyway, so feel free to change those to a different network if you need—thank you!

amritkumarj replied:
pubkeys set in Unstoppable Domains Resolver or Umbra StealthKeyRegistry?

mds1 replied:
pubkeys should now always be set in the Umbra StealthKeyRegistry. It uses compressed keys, so the easiest way to do this is through the UI

mds1 commented:
Hey @amritkumarj, just bumping the last few open comments. The branch is also now out of date with master 🙂

amritkumarj commented:
Hey @mds1 , I have added changes but as rinkeby is discontinued we first need to add goerli in umbra for testing

amritkumarj commented:
@mds1 pushed the changes, please check the cns.test file, as we are using polygon mainnet for testing because rinkeby being discontinued

mds1 commented:
LGTM, thank you @amritkumarj! Do you mind resolving conflicts, then we should be good to merge once CI passes pending any comments from @apbendi. (CI will probably fail now, but should pass once the latest main is incorporated)
